### ISSUE TITLE
Jpeg decompression speeds

### ISSUE THREAD

**Issue by Mencu** [google/neuroglancer#358]
Hello,

after some months of development, we managed to do the following things:

-      decouple the server and the viewer. So, we created a server (still Tornado), that can support multiple viewers (neuroglancer instances) at once.
-     add miscellaneous support for the server, with some already implemented methods from neuroglancer, like link sharing, etc.
-     in order to have flexibility we created our "knossos" protocol. This way, we created our endpoints that transfer the data exactly how we wanted (chunksizes, compression, etc.)
-     because we wanted everything to run smoothly, we configured nginx to directly serve the (compressed) files from the disk. 

Now this is where the problems started to arise: Because we are serving chunks of size 128x128x128 (and on bigger datasets even bigger chunksizes), the decompression step, which now happens in the frontend, gets slow. For this, we've gotten some modules that do it very fastly (example: [@evan/wasm](https://www.npmjs.com/package/@evan/wasm) for snappy decompression).

In the case of the jpeg decompression, we firstly tried the already implemented one from neuroglancer ([jpgjs](https://www.npmjs.com/package/jpeg-js)), but it turns out it is really slow.

Have you tried any other jpeg decompression libraries in the past, except jpgjs? I would be really interested, especially in the WASM ones, so we could speed up the decompression. I have tried [@saschazar/wasm-image-loader](https://www.npmjs.com/package/@saschazar/wasm-image-loader) , but it doesn't seem to build well with neuroglancer.

Would be awesome to have an educated input on this! Thank you!

Best regards,
Andrei Mancu

**Comment by jbms** [comment]
I have not tried a different jpeg implementation, but it seems likely that a wasm one could be faster.  If you are interested in working on that it would be appreciated!  Note though that jpeg decoding should be happening in a separate thread via the async_computation system, or are you not doing that for your Knossos data source?

**Comment by Mencu** [comment]
Yes, the decompression is done inside async_computation(). I will report back if I find a faster working library.

**Comment by william-silversmith** [comment]
I haven't taken it for a test drive, but this looks promising: https://github.com/LinusU/cwasm-jpeg-turbo 

The SIMD support in libjpeg-turbo won't be useful for a while until the support for WASM SIMD is finalized, but I suspect it will still be faster.

**Comment by Mencu** [comment]
Hi,
yes, that's the problem with most WASM jpeg libs, that they can't be used in the web directly.
I'll try to create a web compatible one myself asap.

**Comment by william-silversmith** [comment]
I recently put the bones of this together, but it still needs to be debugged (it can't quite parse the JPEG yet). I might return to fix it later, but if anyone else would like to pick it up in the meantime be my guest. 

https://github.com/william-silversmith/neuroglancer/commit/7e7271864cc3e44c1cf435f3ddee877523c2519d
